libs/net/dal/Services/FileReferenceService.cs

```diff
@@ -254,5 +258,155 @@ public async Task<IEnumerable<FileReference>> GetFiles(DateTime? updatedBefore =
             return Enumerable.Empty<FileReference>();
         }
     }
-    #endregion
+
+    public async Task<Stream> DownloadFromS3Async(string s3Key)
+    {
+        var accessKey = Environment.GetEnvironmentVariable("S3_ACCESS_KEY") ?? throw new InvalidOperationException("S3_ACCESS_KEY environment variable is not set");
+        var secretKey = Environment.GetEnvironmentVariable("S3_SECRET_KEY") ?? throw new InvalidOperationException("S3_SECRET_KEY environment variable is not set");
+        var bucketName = Environment.GetEnvironmentVariable("S3_BUCKET_NAME") ?? throw new InvalidOperationException("S3_BUCKET_NAME environment variable is not set");
+        var serviceUrl = Environment.GetEnvironmentVariable("S3_SERVICE_URL") ?? throw new InvalidOperationException("S3_SERVICE_URL environment variable is not set");
+
+        var config = new AmazonS3Config
+        {
+            ServiceURL = serviceUrl,
+            ForcePathStyle = true
+        };
+
+        using var s3Client = new AmazonS3Client(accessKey, secretKey, config);
+
+        try
+        {
+            var request = new GetObjectRequest
+            {
+                BucketName = bucketName,
+                Key = s3Key
+            };
+
+            var response = await s3Client.GetObjectAsync(request);
+
+            if (response.HttpStatusCode == System.Net.HttpStatusCode.OK)
+            {
+                var memoryStream = new MemoryStream();
+                await response.ResponseStream.CopyToAsync(memoryStream);
+                memoryStream.Position = 0;
+                return memoryStream;
+            }
+            else
+            {
+                Logger.LogError("Failed to download file from S3: {S3Key}", s3Key);
+                throw new Exception($"Failed to download file from S3: {s3Key}");
+            }
+        }
+        catch (AmazonS3Exception ex)
+        {
+            Logger.LogError(ex, "Error retrieving file from S3: {S3Key}", s3Key);
+            throw;
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError(ex, "Unexpected error when downloading file from S3: {S3Key}", s3Key);
+            throw;
+        }
+    }
+
+    public async Task<FileReference?> GetByS3PathAsync(string s3Path)
+    {
+        return await this.Context.FileReferences
```

Comment:
You shouldn't be searching the database for a file reference by path.  You should only be searching for file reference by its primary key.